### PR TITLE
Test time augmentations

### PR DESCRIPTION
**Pull request by edyoshikun**
This PR adds the test time augmentations to the prediction steps.
These are currently hardcoded to be rotations of (0,90,180 and 270 deg) by setting the `trainer.test_time_augmentation=True`

Ideally, we want this to be added during training the paper suggests so that we are rotationally invariant.

### CODE REVIEW

File: viscy/data/hcs.py [hunk 1]
```diff
@@ -434,7 +434,10 @@ def _setup_test(self, dataset_settings: dict):
                 **dataset_settings,
             )
 
-    def _setup_predict(self, dataset_settings: dict):
+    def _setup_predict(
+        self,
+        dataset_settings: dict,
```
Comment (ziw-liu) [inline viscy/data/hcs.py:439]: Extra comma

File: viscy/light/engine.py [hunk 2]
```diff
@@ -113,6 +113,10 @@ class VSUNet(LightningModule):
     :param bool test_evaluate_cellpose:
         evaluate the performance of the CellPose model instead of the trained model
         in test stage, defaults to False
+    :param bool test_time_augmentations:
+        apply test time augmentations in test stage, defaults to False
+    :param Literal['mean', 'median'] tta_type:
```
Comment (ziw-liu) [inline viscy/light/engine.py:118]: The union members are not consistent with the signature.

File: viscy/light/engine.py [hunk 3]
```diff
@@ -315,8 +324,60 @@ def _log_segmentation_metrics(
...
+        source = batch["source"]
+
+        if self.test_time_augmentations:
+            # Save the yx coords to crop post rotations
+            self._original_shape_yx = source.shape[-2:]
+
+            predictions = []
+            # for aug, de_aug in zip(
+            #     test_time_augmentations_list, test_time_augmentations_list_inv
+            # ):
```
Comment (ziw-liu) [inline viscy/light/engine.py:336]: What does this do?

File: viscy/light/engine.py [hunk 4]
```diff
@@ -315,8 +324,60 @@ def _log_segmentation_metrics(
...
+            for i in range(4):
+                augmented = self._rotate_volume(source, k=i, spatial_axes=(1, 2))
+                augmented = self._predict_pad(augmented)
+                # predict and move to CPU
+                augmented_prediction = self.forward(augmented)
```
Comment (ziw-liu) [inline viscy/light/engine.py:341]: Why do inference on CPU? This is going to be slow.

File: viscy/light/engine.py [hunk 5]
```diff
@@ -315,8 +324,60 @@ def _log_segmentation_metrics(
...
+                )
+
+                # Undo rotation and padding
+                predictions.append(de_augmented_prediction.cpu())
+                del (
+                    augmented,
+                    augmented_prediction,
+                    de_augmented_prediction,
+                )
+                torch.cuda.empty_cache()  #
```
Comment (ziw-liu) [inline viscy/light/engine.py:359]: Why clear CUDA cache when things are all on the CPU?

File: viscy/light/engine.py [hunk 6]
```diff
@@ -315,8 +324,60 @@ def _log_segmentation_metrics(
         )
 
     def predict_step(self, batch: Sample, batch_idx: int, dataloader_idx: int = 0):
-        source = self._predict_pad(batch["source"])
-        return self._predict_pad.inverse(self.forward(source))
+        source = batch["source"]
+
+        if self.test_time_augmentations:
```
Comment (ziw-liu) [inline viscy/light/engine.py:329]: This branch is very long. It could be another method.

File: viscy/light/predict_writer.py [hunk 7]
```diff
@@ -32,7 +32,31 @@ def _resize_image(image: ImageArray, t_index: int, z_slice: slice) -> None:
         )
 
 
-def _blend_in(old_stack: NDArray, new_stack: NDArray, z_slice: slice) -> NDArray:
+def _blend_in(old_stack: NDArray, new_stack: NDArray, z_slice: slice) -> None:
```
Comment (ziw-liu) [inline viscy/light/predict_writer.py:35]: This function does return an NDArray, why change to None?

**Comment by ziw-liu** [comment]
@edyoshikun can you clean up the diff?

**Comment by edyoshikun** [comment]
I need to figure out why when I merge main to my local it always does this thing... Thanks. Ill clean it up @ziw-liu

**Review by ziw-liu — APPROVED** [review]

**Pull request merged**